apiserver: new `/actions/reboot` endpoint

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
N/A


**Description of changes:**
Adds new API path: `/actions/reboot` for initiating reboots.



**Testing done:**
Built images, created AMI, launched instance with said AMI
Was able to reboot the machine from both the control container and admin container with:
`apiclient -v -u /actions/reboot -m POST`. 


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
apiserver: new `/actions/reboot` endpoint

Adds new API path: `/actions/reboot` for initiating reboots.

tjkirch commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -352,6 +353,27 @@ async fn get_configuration_files(
+    // Currently relying on systemd to add a relay before the reboot.
+    ensure!(
+        Command::new("/bin/systemd-run")
+            // Add a 10 second delay to shutdown
```
What's the reasoning for a 10 second delay?  Service ordering constraints should take care of the logical flow of shutdown, so if the only thing we care about here is having enough time to return an HTTP response code, we only need a tiny fraction of this.  Did you test and find that we do require a delay at all?

etungsten replied:
The 10 seconds is arbitrary and I'm OK with changing it to any value that makes sense. I haven't tested directly calling `shutdown -r`, testing that now.

tjkirch commented on sources/api/openapi.yaml:
```diff
@@ -314,3 +314,13 @@ paths:
+      operationId: "reboot"
+      responses:
+        204:
+          description: "Successfully requested - reboot in 10 seconds"
```
```suggestion
          description: "Reboot requested"
```
Let's not include the time here; the time could change and I think there's no need to remember to include it here.

tjkirch commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -352,6 +353,27 @@ async fn get_configuration_files(
+            .arg("-r")
+            .status()
+            .context(error::SystemdRun)?
+            .success(),
+        error::Reboot
```
There are a lot of reasons this oculd happen, so I think we should include the stderr output and return code if the command fails, which would mean using `output` instead of `success` and the simple `ensure`.

etungsten commented:
Addresses @tjkirch 's comments.
* Removed the arbitrary delay.
* Directly call `/sbin/shutdown` instead of going through systemd
* Added more information to the `Reboot` error.

Tested with multiple instance and I am able to get a reponse status back every time.

etungsten commented:
Push above renames `SystemdRun` error to `Shutdown`.

zmrow reviewed (approved):
🔩

tjkirch commented on sources/api/apiserver/src/server/error.rs:
```diff
@@ -115,6 +115,12 @@ pub enum Error {
+    #[snafu(display("Unable to start shutdown: {}", source))]
+    Shutdown { source: io::Error },
+
+    #[snafu(display("Failed to reboot, exit code: {}, stderr: {:?}", exit_code, stderr))]
```
```suggestion
    #[snafu(display("Failed to reboot, exit code: {}, stderr: {}", exit_code, stderr.from_utf8_lossy()))]
```
I think you'd see `Vec[104, 105]` without turning it into a string.

tjkirch commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -352,6 +354,27 @@ async fn get_configuration_files(
+        error::Reboot {
+            exit_code: match output.status.code() {
+                Some(code) => code,
+                None => output.status.signal().unwrap_or_else(|| 1),
```
```suggestion
                None => output.status.signal().unwrap_or(1),
```
Shouldn't need the closure for a Copy type like an integer that's trivial to construct.

tjkirch reviewed (approved):
Just nits

etungsten commented:
Addresses @tjkirch 's comments.
